// @textea/json-viewer/dist/out/stores/JsonViewerStore.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.useJsonViewerStore = exports.JsonViewerProvider = exports.JsonViewerStoreContext = exports.createJsonViewerStore = void 0;
var react_1 = require("react");
var zustand_1 = require("zustand");
var base16_1 = require("../theme/base16");
var DefaultKeyRenderer = function () { return null; };
DefaultKeyRenderer.when = function () { return false; };
var createJsonViewerStore = function (props) {
    return (0, zustand_1.create)()(function (set, get) {
        var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m, _o, _p, _q, _r, _s;
        return ({
            // provided by user
            enableClipboard: (_a = props.enableClipboard) !== null && _a !== void 0 ? _a : true,
            highlightUpdates: (_b = props.highlightUpdates) !== null && _b !== void 0 ? _b : false,
            indentWidth: (_c = props.indentWidth) !== null && _c !== void 0 ? _c : 3,
            groupArraysAfterLength: (_d = props.groupArraysAfterLength) !== null && _d !== void 0 ? _d : 100,
            collapseStringsAfterLength: (props.collapseStringsAfterLength === false)
                ? Number.MAX_VALUE
                : (_e = props.collapseStringsAfterLength) !== null && _e !== void 0 ? _e : 50,
            maxDisplayLength: (_f = props.maxDisplayLength) !== null && _f !== void 0 ? _f : 30,
            rootName: (_g = props.rootName) !== null && _g !== void 0 ? _g : 'root',
            onChange: (_h = props.onChange) !== null && _h !== void 0 ? _h : (function () { }),
            onCopy: (_j = props.onCopy) !== null && _j !== void 0 ? _j : undefined,
            onSelect: (_k = props.onSelect) !== null && _k !== void 0 ? _k : undefined,
            keyRenderer: (_l = props.keyRenderer) !== null && _l !== void 0 ? _l : DefaultKeyRenderer,
            editable: (_m = props.editable) !== null && _m !== void 0 ? _m : false,
            defaultInspectDepth: (_o = props.defaultInspectDepth) !== null && _o !== void 0 ? _o : 5,
            objectSortKeys: (_p = props.objectSortKeys) !== null && _p !== void 0 ? _p : false,
            quotesOnKeys: (_q = props.quotesOnKeys) !== null && _q !== void 0 ? _q : true,
            displayDataTypes: (_r = props.displayDataTypes) !== null && _r !== void 0 ? _r : true,
            // internal state
            inspectCache: {},
            hoverPath: null,
            colorspace: base16_1.lightColorspace,
            value: props.value,
            prevValue: undefined,
            displayObjectSize: (_s = props.displayObjectSize) !== null && _s !== void 0 ? _s : true,
            getInspectCache: function (path, nestedIndex) {
                var target = nestedIndex !== undefined
                    ? path.join('.') +
                        "[".concat(nestedIndex, "]nt")
                    : path.join('.');
                return get().inspectCache[target];
            },
            setInspectCache: function (path, action, nestedIndex) {
                var target = nestedIndex !== undefined
                    ? path.join('.') +
                        "[".concat(nestedIndex, "]nt")
                    : path.join('.');
                set(function (state) {
                    var _a;
                    return ({
                        inspectCache: __assign(__assign({}, state.inspectCache), (_a = {}, _a[target] = typeof action === 'function'
                            ? action(state.inspectCache[target])
                            : action, _a))
                    });
                });
            },
            setHover: function (path, nestedIndex) {
                set({
                    hoverPath: path
                        ? ({ path: path, nestedIndex: nestedIndex })
                        : null
                });
            }
        });
    });
};
exports.createJsonViewerStore = createJsonViewerStore;
// @ts-expect-error we intentionally want to pass undefined to the context
// See https://github.com/DefinitelyTyped/DefinitelyTyped/pull/24509#issuecomment-382213106
exports.JsonViewerStoreContext = (0, react_1.createContext)(undefined);
exports.JsonViewerProvider = exports.JsonViewerStoreContext.Provider;
var useJsonViewerStore = function (selector, equalityFn) {
    var store = (0, react_1.useContext)(exports.JsonViewerStoreContext);
    return (0, zustand_1.useStore)(store, selector, equalityFn);
};
exports.useJsonViewerStore = useJsonViewerStore;
